src/hidden.rs

```diff
@@ -20,74 +20,108 @@
 // WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-//! A wrapper to conceal secrets when output into logs or displayed.
+//! A generic type that hides data from being displayed or written, keeps it on the heap, zeroizes it when it goes away,
+//! limits access, and allows for type enforcement.
 
-use std::fmt;
+use std::{any::type_name, fmt, marker::PhantomData};
 
 use serde::{Deserialize, Serialize};
+use zeroize::Zeroize;
 
-/// A simple struct with a single inner value to wrap content of any type.
-#[derive(Copy, Clone, Serialize, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+/// A marker trait for labeling different hidden types
+pub trait HiddenLabel {}
+
+/// Create a hidden type label
+#[macro_export]
+macro_rules! hidden_label {
+    ($name:ident) => {
+        #[derive(Clone, Debug, Default, Eq, Ord, PartialEq, PartialOrd)]
+        pub struct $name;
+
+        impl HiddenLabel for $name {}
+    };
+}
+
+// A default hidden type label; only use this if you're absolutely sure you don't care about type enforcement
```

Comment:
I would add more docs to what we mean by type enforcement. I think it is important to be explicit that the main goal of type enforcement is to not reuse `Hidden` instances across different contexts.

Reply:
Yeah, if this design is approved, I'll add much more complete documentation.